### ISSUE TITLE
2X rendering/2X fetch calls on page revalidation ISR [production build] - Next.js v14.0.4-canary.15

### ISSUE THREAD

**Issue by leoBergerot** [vercel/next.js#58736]
### Link to the code that reproduces this issue
https://codesandbox.io/p/devbox/naughty-dream-4kmzfm?file=%2Fapp%2Fpage.tsx%3A11%2C1

### To Reproduce

Hello,

I am experiencing an issue with double fetch calls during the page revalidation process in a Next.js 14 production environment. This issue occurs when navigating back to a page that is undergoing revalidation.

**Description**:
- The issue occurs in production, not in development mode.
- I am using Next.js 14 with Incremental Static Regeneration (ISR).
- The problem occur on dynamic segment and normal segment
- page revalidate = 30
- fetch revalidate = 30

**Example**:
When a page is revalidated with single request, I see the following behavior in my server logs:
<img width="877" alt="Capture d’écran 2023-11-21 à 16 21 45" src="https://github.com/vercel/next.js/assets/24389962/4ee20187-b099-4028-a3ce-089f6a21ca6e">

This indicates that the fetch function for retrieving posts data is being called twice. And the cached data was beeing update twice.

on my server api log: 
<img width="626" alt="Capture d’écran 2023-11-21 à 16 22 51" src="https://github.com/vercel/next.js/assets/24389962/a8f3d2fd-139c-4cb6-ab9c-d189cc41ee0c">
This indicates that the fetch function for retrieving posts data is really called twice.

### Current vs. Expected behavior

I would expect the fetch call to only occur once during the revalidation process, not twice as currently observed.

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: linux
  Arch: arm64
  Version: #1 SMP PREEMPT Wed Oct 25 16:32:24 UTC 2023
Binaries:
  Node: 18.18.2
  npm: 9.8.1
  Yarn: 1.22.19
  pnpm: N/A
Relevant Packages:
  next: 14.0.4-canary.4
  eslint-config-next: 13.5.6
  react: 18.2.0
  react-dom: 18.2.0
  typescript: N/A
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Not sure, Data fetching (gS(S)P, getInitialProps)

### Additional context

Additionally, I am using a file system-based cache in my application.  isrMemoryCacheSize: 0
after updating to 14.0.4-canary.2, the revalidation issues would be resolved. However, the double fetch calls during revalidation are still a concern.

<sub>[NEXT-3233](https://linear.app/vercel/issue/NEXT-3233/2x-rendering2x-fetch-calls-on-page-revalidation-isr-[production-build])</sub>

**Comment by leoBergerot** [comment]
I've encountered an issue that I'm not sure is normal. From my perspective, it seems highly irregular, especially in the context of fetch operations as demonstrated here: https://codesandbox.io/p/devbox/naughty-dream-4kmzfm?file=%2Fapp%2Fpage.tsx%3A11%2C1. The issue leads to double the resource consumption than expected. This could have significant repercussions since the data retrieved and displayed might not be the latest. This becomes critical in applications where up-to-date data is essential.

I've updated my example using the API https://catfact.ninja/fact, which returns different data on each call. This demonstrates that page revalidation results in two API calls instead of one. I've logged this behavior to highlight the problem.

I am seeking feedback from other users who might be facing a similar issue. In scenarios where multiple API calls are integral to constructing a page, this issue could lead to substantial resource consumption, effectively doubling the expected usage. I find this quite concerning and hope to find a solution or workaround.

Any insights or suggestions from the community would be greatly appreciated.

Thank you for your attention to this matter.

**Comment by OlegLustenko** [comment]
Check this with production build

**Comment by modeseven** [comment]
Can confirm the issue, I'm on ```14.0.3``` . it fires twice on build as well. My app will use this feature for 100s of pages and doubling the queries it does is really a bummer. Anyone know of a workaround?

**Comment by OlegLustenko** [comment]
@leoBergerot check this version https://codesandbox.io/p/devbox/silly-snow-ql9kd9?file=%2Fapp%2Fpage.tsx%3A22%2C13

You need to build and open the link that is generated with `npm run start`.

It renders twice every 5 seconds(revalidate 5 sec), however, there is only one fetch.

With `export const dynamic = "force-dynamic"` it's 2 re-renders on a first-page load, but with a browser refresh it's only 1 re-render.

Hope it helps satisfy your needs

**Comment by leoBergerot** [comment]
[OlegLustenko](https://github.com/OlegLustenko) Hi, thank you for your interest in the topic. However, in your sample, there are duplicated queries. Please check the console log on start tab, where you can see that the 'catfac' object is not the same at the (i'm update your sample) beginninghttps://codesandbox.io/p/devbox/summer-shadow-m6pw2z?file=%2Fapp%2Fpage.tsx%3A16%2C34

**Comment by leoBergerot** [comment]
> bummer



> Can confirm the issue, I'm on `14.0.3` . it fires twice on build as well. My app will use this feature for 100s of pages and doubling the queries it does is really a bummer. Anyone know of a workaround?

I disappointed too, it's huge consumption of resources for anything…, I am attempting to debug the incremental cache in 'next/dist/server/lib/incremental-cache/index.js,' but the issue does not originate from this location. It occurs earlier in the process about me

**Comment by OlegLustenko** [comment]
Potentially will be fixed with this PR

https://github.com/vercel/next.js/pull/58990

**Comment by FerrariAndrea** [comment]
Hi everyone, I think I have the same issue, I'm using some dynamic page using: generateStaticParams.
Tested with Next.js 14.0.3 and last canary: 14.0.4-canary.2
Same scenario, evry page is running my fetch (that is my DB query) 4 time.
And I'm expecting to see it only 2 times, once for the real page build and the second one because it is used inside "generateMetadata".

So consider the page using a slag, and consider "fetch(Q)" as a function that wraps the next.js fetch  adding logs on the time spent running the request:
```
export async function generateMetadata({ params }){... fetch(Q) ...}
export async function generateStaticParams() {...}
export default async function NameOfThePage({ params }){... fetch(Q)...}
```

My output for evry page prebuild (during "next build") is:

```
TIMING NameOfThePage[slug]: 1496ms
TIMING NameOfThePage[slug]: 1497ms
TIMING NameOfThePage[slug]: 1495ms
TIMING NameOfThePage[slug]: 1493ms
```

So evry page seems to be built twice.
Another consideration is the time required to run the fetch is the same, which means the fetch cache is ignored, I think this is another issue, and I'm expecting to see the timing of the last 3 requests really less than the first one.

For other information, I'm using the "{ revalidate: revalidate }" option inside the next.js fetch, like:
```
fetch(URL,{ revalidate: 3600})
```


Not sure if I'm going off-topic, but this is the more similar issue that I found.

Thx

**Comment by yanghuilong** [comment]
I also encountered the same problem.
![image](https://github.com/vercel/next.js/assets/10112138/433f8c58-a4ba-44b9-929a-deab2a0985de)

**Comment by leoBergerot** [comment]
@OlegLustenko The problem is already the same with https://github.com/vercel/next.js/releases/tag/v14.0.4-canary.22 :/
I update the sample with this release -> https://codesandbox.io/p/devbox/naughty-dream-4kmzfm?file=%2F.codesandbox%2Ftasks.json%3A26%2C37

**Comment by leoBergerot** [comment]
Hi, Same problem on 14.0.4 -> https://codesandbox.io/p/devbox/laughing-allen-z3plg7

**Comment by JoeyKhd** [comment]
Can vouch, currently on canary.59 and the same problem

**Comment by varzy** [comment]
Same problem on 14.2.0-canary.23. I think this is a very serious problem, is there any temporary solution?

**Comment by leoBergerot** [comment]
No, I'm very disappointed. I have a lack of performance and it takes so many resources for anything. So, I've put a cache on my endpoint in order to avoid wasting resources by making revalidation calls twice.

**Comment by varzy** [comment]
> No, I'm very disappointed. I have a lack of performance and it takes so many resources for anything. So, I've put a cache on my endpoint in order to avoid wasting resources by making revalidation calls twice.

Same with me. I had to remove my expensive fetching from the page component to avoid this issue, and then I provided an api to manually trigger my expensive fetching after build.
It works, but it's still quite painful 😣.

**Comment by Smirow** [comment]
Same issue using ISR with SvelteKit, so probably more an issue with Vercel/Build output API.

**Comment by mdahiemstra** [comment]
We are also experiencing this problem and Im very surprised i did not notice it earlier. I did not notice until I added the vercel/otel package to our instance and started investigating some performance issues. I see all the fetch requests twice in my Jaeger UI. 

The thing is, its "exactly" the same, so I doubt the request is actually done twice, but its logged twice. So I'm not sure it adds any latency but it certainly is very confusing and cluttering.

**Comment by feedthejim** [comment]
We're gonna take a look at this this week, sorry for the delay on this!

**Comment by mdahiemstra** [comment]
Let me know if i can do anything to help to speed up the processes.

**Comment by josegutierro** [comment]
Exact the same issue here

**Comment by jbeyermonday** [comment]
Hi, is there any news on this or did anybody find a workaround? 
We are observing the exact same behavior and are struggling anyway with request to a content headless server. Now with the 2x rendering we practically doubling our requests and presumably a lot of unnecesary extra computing we have to sustain.

**Comment by ztanner** [comment]
Hey folks -- thank you for reporting this. It looks like this is because we currently walk the component tree twice when statically generating the page (so you'd also observe this behavior during a regular `next build`). 

- The first call starts here -- during static generation:
https://github.com/vercel/next.js/blob/270a9db0567808db5d53e7a18c7be0a4710796e8/packages/next/src/server/app-render/app-render.tsx#L833-L835

- `generateFlight` invokes `walkTreeWithFlightRouterState` function, https://github.com/vercel/next.js/blob/270a9db0567808db5d53e7a18c7be0a4710796e8/packages/next/src/server/app-render/app-render.tsx#L310-L330
- this function then calls `createComponentTree` internally: https://github.com/vercel/next.js/blob/270a9db0567808db5d53e7a18c7be0a4710796e8/packages/next/src/server/app-render/walk-tree-with-flight-router-state.tsx#L142-L158

- The second call happens here:
https://github.com/vercel/next.js/blob/270a9db0567808db5d53e7a18c7be0a4710796e8/packages/next/src/server/app-render/app-render.tsx#L435-L448

This is on our radar and something we'll be addressing -- ideally we should be able to do this in a single pass and not require two invocations.

**Comment by enchorb** [comment]
This is such a broken framework, how could something like this make it to release. Was fine in 14.1.4 but occurring in 14.2.0+ for us. 

Server action revalidation causes 2 requests, however after browser reloading the page it goes back to the expected 1 request.

**Comment by mspoerer** [comment]
Dear Next.js Team,

We kindly request that you address the issue of double fetch calls during page revalidation (Issue #58736) as soon as possible. This problem is causing our resource usage and costs to double, significantly impacting our ability to use the framework efficiently. We appreciate your hard work and hope for a swift resolution.

**Comment by beyeja** [comment]
I can attest to that. 
Additionally it seems to affect non memory caching quite a bit as it even double accesses data-caches. We log our access times when fetching and every double cache hit measures first a quick and then a slower access performance. This is probably no issue when reading from memory but file system or any remote cache can simply not benefit from accessing a data cache entry simultaneously. We rely strongly on a shared cache.

Is there any info you can give when this will be addressed or any temporary workaround idea?

**Comment by baba43** [comment]
This and https://github.com/vercel/next.js/discussions/48481 make the so promised features horrible to use. I really do not want to be rude or ungrateful but why are this pain points still there? I also checked Next.js 15 RC and both issues are still there.

**Comment by feedthejim** [comment]
Sorry again for the delay. We uncovered the root of this issue, and we're looking at optimizing this soon.

**Comment by PramodTKodag** [comment]
Any update on this ?

**Comment by Cha195** [comment]
Hi @ztanner. A Nexjs beginner here. Can you give some more information on this statement.
> It looks like this is because we currently walk the component tree twice when statically generating the page 

What do you mean by "walking" the component tree? Why is there a need to walk the component tree twice while statically generating pages?
Eager to know more.

Thanks

**Comment by ztanner** [comment]
> Hi @ztanner. A Nextjs beginner here. Can you give some more information on this statement.
> 
> > It looks like this is because we currently walk the component tree twice when statically generating the page
> 
> What do you mean by "walking" the component tree? Why is there a need to walk the component tree twice while statically generating pages? Eager to know more.
> 
> Thanks

Hey @Cha195, sorry that I missed this question. I added some more information in the PR that fixes it (#67680). By "walking" the component tree, I meant "rendering". 

For static pages, Next.js creates two files: an HTML file with the rendered page, as well as an RSC payload. This RSC payload is used by the router & React during client navigations and contains information about the components that should be rendered for the given page.

Previously these separate files were created in separate renders. The refactor allows us to create them in a single render.
